torchmetrics/metric.py

```diff
@@ -206,6 +208,15 @@ def _wrap_compute(self, compute):
 
         @functools.wraps(compute)
         def wrapped_func(*args, **kwargs):
+            if not self._update_called:
+                rank_zero_warn(
+                    f"The ``compute`` method of metric {self.__class__.__name__}"
+                    " was called before the ``update`` method which may lead to errors,"
+                    " as metric states have not yet been updated. Will return 0.0 instead.",
+                    UserWarning
+                )
+                return 0.0
```

Comment:
shall this be rather NaN?

Reply:
Fine with me :]
What does lightning do in `self.log` if a `NaN` is passed in?

Reply:
@tchaton ^^

Reply:
Great question. Looks like Lightning doesn't handle NaN values. It will store it and then produce a NaN value at reduction time.
@SkafteNicki Can you have a look ? I think we might want to skip saving the result object is NaN is logged.

Reply:
And 0 would actually make the reduction False.

Reply:
let me try to update this on the lightning side

Reply:
I'm not sure if it's worth it to return 0.0 - some metrics are returning tuples/vectors when computed correctly and there will be code to handle it, in that case, there will be a weird exception down the line due to type-mismatch. Might be better to just raise instead.

In terms of type-mismatch, this return 0 will also break type-checking if it was smart enough to understand all the dynamic wrapping happening in the first place :)

Reply:
@maximsch2 so you propose that we raise an error instead?

Reply:
coming back, how about the NaN instead of 0?

Reply:
The question is - do we want the training to proceed and do we expect this to happen deep into the training? 

Given that most of the time we will probably get the error right away (on_step=True in training_step and validation will be caught by sanity checking), I think we should throw an exception here. Alternatively - maybe just warn and do nothing? Let `compute` fail if it does? 

Another aspect to think about - say I have a metric that is stateless and the only thing it does is that in `compute()` it queries RAM utilization or CPU load or something else and returns it. Given that it's going to be rare and probably done by more advanced users, asking them to call update() can be fair game I guess.

Reply:
then lets settle for raising an warning to the users. Then it depends on the metric what will happen afterwards:
* metrics using list states will fail because `compute` tries to call `torch.cat` on empty list
* metrics of the form `num/denom` will give `nan` as it in most cases it will be `0/0`
* some metrics will behave nicely (like confusionmatrix with no normalization) because the metric is well defined from the initial state